### PR TITLE
[Component] Leo Icon

### LINKED ISSUE #45: Icons
### Serving / Requesting
We want to allow:
- Consumers to self-host all the .svg files so that they can reduce duplication among bundles (e.g. brave-core serving icons from chrome://resources/leo-icons/x.svg
  - Consumers could choose to only serve .svg files they are using, to conserve package size
  - Consumers could serve icons from a central URL but Leo JS from local bundles
- Consumers can define a custom library of .svg files, and get the benefits of a `LeoIconComponent`

All icons exist in repo as separate .svg files, e.g. `/leo/icons/xyz.svg`
`LeoIconComponent` fetches the file and renders the svg, adds css class to it.
Consumer passes in icon-name and 
```
<LeoIcon library="leo" label="" name="search" />
````

Take inspiration from https://shoelace.style/components/icon / https://github.com/shoelace-style/shoelace/tree/next/src/components/icon

### Benefits of `LeoIconComponent`
- Change fill / stroke to `currentColor` so it inherits `color` via css
- Restrict width and height to `--leo-icon-size` so that the icon fits in to containers, e.g. different button types
- Fetches and inlines

### What it won't do
- Accept custom color property, this should be cascaded through css, i.e. set `color: pink` on the custom element or container or anywhere. Same with `--leo-icon-size`.

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Fixes #45 

Storybook: https://leo-storybook-git-leo-icon-fallaciousreasoning.vercel.app/?path=/story/icon--default

This adds a `leo-icon` component. Currently the following is supported:

- Set the color via `color: red`
- Don't override color of `-color` icons (this relies on the naming convention remaining the same)
- Set size via `--leo-icon-size`

In addition, this package adds support for downloading all the icons from Figma via the `npm run update-icons` command, using a fork of https://github.com/slawomirkolodziej/figma-api-exporter

~~However, the licensing for that package is unclear. I've requested clarification (https://github.com/slawomirkolodziej/figma-api-exporter/issues/5) and if they don't add a license we might have to roll our own, which would be a bit sad but not too difficult.~~ ~~The author is pretty responsive, and I think it's going to be MIT licensed!~~ The package is MIT licensed!

Note: We're pointing at a revision on the Brave fork of the library for security: https://github.com/brave/security/issues/1137#issuecomment-1341769048 

[Screencast from 2022-11-23 12-16-53.webm](https://user-images.githubusercontent.com/7678024/203440441-2e5da95e-0a4e-4f7e-ac02-2bb58af004c5.webm)

### CODE REVIEW

**Comment by petemill** [comment]
@AlanBreck and I were talking about the colors and whilst the mutate function is great to have, perhaps it would be more efficient (as well as allow some non-js implementations) if we stripped the fill attributes in the update-icons.js script?

**Comment by fallaciousreasoning** [comment]
Also flags don't have the color suffix so we should special case them (and possibly rename them so we can render the icons with `Country=NZ` instead of `Country=NZ New Zealand (Aotearoa)`

**Comment by fallaciousreasoning** [comment]
Okay, I've pushed the following changes:
- Mutate the icons when we download them instead of at runtime.
- Country Icons are always treated as color icons
- Added support for icons via a slot, as well as via the name attribute, which solves @petemill's `import` problem.
- Added a story for custom icons

File: package.json
Comment (AlanBreck) [inline package.json:37]: Do we need a sec review on this one?

File: web-components/icon/Icon.svelte
Comment (AlanBreck) [inline web-components/icon/Icon.svelte:3]: Curiosity more than anything: why is this code in `context="module"` and not the main `script` tag?

File: web-components/icon/Icon.svelte
Comment (AlanBreck) [inline web-components/icon/Icon.svelte:21]: I haven't seen this syntax before! How does this work?

File: web-components/icon/library.default.ts
Comment (AlanBreck) [inline web-components/icon/library.default.ts:4]: How is this meant to be used?

**Review by AlanBreck — COMMENTED** [review]
Sorry for some of the curiosity questions. Two more questions, though:
1. Can we make the name case insensitive? E.g. `ATM` and `atm` both work?
2. Setting `library` knob in Storybook throws an error.

File: package.json
Comment (fallaciousreasoning) [reply package.json:37]: We do indeed - it's over here https://github.com/brave/security/issues/1137
Includes the dependency & the `@html` review :smile:

File: web-components/icon/Icon.svelte
Comment (fallaciousreasoning) [reply web-components/icon/Icon.svelte:21]: It's null propagation - essentially only call the method if it's not null

`ts
const f = foo?.bar?.baz

// is the same as
const f = foo
	? foo.bar
		? foo.bar.baz
        : undefined
    : undefined

// you can also use it for arrays/funcs but the syntax is a bit more funky

const nthItem = items?.[n]
// which is the same as 
const nthItem = items ? items[n] : undefined

// and for functions, they'll only be called if they're defined
const result = func?.(stuff, things)

// which is
const result = func ? func(stuff, things) : undefined

File: web-components/icon/library.default.ts
Comment (fallaciousreasoning) [reply web-components/icon/library.default.ts:4]: Basically, you can use it to set where the icons are served from. I.e. if we wanted to serve them from `chrome://resources/leo-icons` we could call `setDefaultIconsPath('chrome://resources/leo-icons')` and it would fetch the `atm` icon from `chrome://resources/leo-icons/atm.svg`

File: web-components/icon/Icon.svelte
Comment (fallaciousreasoning) [reply web-components/icon/Icon.svelte:3]: So that's so the cache is shared between instances of the icon component. If it was in the main script tag each icon would fetch the SVG again.

**Comment by fallaciousreasoning** [comment]
> Setting library knob in Storybook throws an error.

FWIW, that's expected - only the default library is defined and trying to use a non-existent library probably should break, as it wouldn't display anything, so is probably a mistake.

> Can we make the name case insensitive? E.g. ATM and atm both work?

We definitely can - we could lowercase all icon names. I was thinking of creating a typescript file with all the leo icons in it though, so we get intellisense. No strong feelings on caseSeNsiTivitY though. @petemill do you have a preference?

File: web-components/icon/Icon.svelte
Comment (AlanBreck) [reply web-components/icon/Icon.svelte:21]: Super cool.

**Review by AlanBreck — APPROVED** [review]

**Comment by fallaciousreasoning** [comment]
Currently, we have a few actions to make for the Security Review:
- Add DOMPurify to the icon component.
- Depend on a fork of https://github.com/slawomirkolodziej/figma-api-exporter to guard against potentially malicious changes.

File: web-components/icon/icon.svelte
Comment (petemill) [inline web-components/icon/icon.svelte:43]: How about we use `background-color: currentColor` and then we can just rely on color cascading for both inline and background variants?

Comment (fallaciousreasoning) [reply web-components/icon/icon.svelte:43]: I had no idea you could do this :O I thought `currentColor` was a special SVG thing ! Awesome!

**Review by thypon — APPROVED** [review]

**Review by AlanBreck — APPROVED** [review]

File: package.json
Comment (petemill) [inline package.json:11]: Everything else in dependencies has the version pinned, so seems like we should continue to do that for these new ones

**Review by petemill — APPROVED** [review]

**Comment by fallaciousreasoning** [comment]
I'm holding off on the merge till the bundling PR #161 lands, as it has priority, and this is huge. (sorry for all the force pushes, I accidentally push the bundling PR to the icons branch :fearful: )

**Pull request merged**
